### ISSUE TITLE
Segfault when using external search engine

### ISSUE THREAD

**Issue by Schulteatq** [doxygen/doxygen#8757]
**Describe the bug**
When running Doxygen on any non-empty code base with `EXTERNAL_SEARCH = YES`, it will crash/segfault. Setting `EXTERNAL_SEARCH` to `NO` will work as expected, however, the searchdata.xml file is not there obvioulsy.

Screenshot from the stack trace:
![grafik](https://user-images.githubusercontent.com/6652808/131866539-f7278465-243a-4edb-afdc-662a445390c8.png)

Inspecting `m_str` shows a memory value of `DDDDDDDD` indicating that this location was already freed, which in return will trigger the segfault when calling `realloc()` on that address. Also the debug heap reports "Invalid address specified to RtlValidateHeap( 000002235C090000, 000002235C1266E0 )".
I'm quite confident that the `GrowBuf` copy/move-assignment operators and copy/move ctors introduced in [this commit](https://github.com/doxygen/doxygen/commit/c48639744a6fa118b9851b307107994ba93ce4c8#diff-023d9280962670f95c4ea4f93517276f15cddc46fb9dcc2311a0e4ea88967b82) are broken since they do not free/reset the original `m_str` member, yielding duplicate owership of m_str.

Unfortunately, I'm was not able to make Doxygen build on my Windows machine (lack of dependencies) so I can not fully confirm my assumption by compiling a potential fix. Still I'm confident that this is the root cause since elements of this type are stored by value in a std container.

**To Reproduce**
Run doxygen on any non-empty code base with `EXTERNAL_SEARCH = YES` set.

**Version**
Happens with the 1.9.2 Windows binaries from the website. Same behavior with the latest Windows binaries from Github Actions  (version 1.9.3 (c9fc3ac4b9c83a9b0ed853b4508bc9f604ac7deb*)). 
Version 1.9.1 was still fine, IIRC.

**Comment by Schulteatq** [comment]
@doxygen Sorry that my initial report was not sufficient for reproduction. Here is a minimal example: 
[GrowBuf Crash.zip](https://github.com/doxygen/doxygen/files/7101733/GrowBuf.Crash.zip).
This crashes on my machine with `Doxygen version used: 1.9.3 (c9fc3ac4b9c83a9b0ed853b4508bc9f604ac7deb*)`, Debug version pulled from the corresponding Github Actions page. 
Same thing happens with the 1.9.2 release from the official website, just less noisy due to the absence of a debug heap.

Regarding `GrowBuf`: Copy ctor and copy assignment are indeed ok except for the memory leak in the copy assignment that you already mentioned. The move ctor and move assignment however are broken: All three members of `GrowBuf` are primitive types, therefore a `std::move()` on them does not move anything but is effectively a no-op so that the values are copied. As a result, at the end of either function ´other.m_str` and `this->m_str` point to the exact same address both owning it. Once the moved-from object gets deleted it will free `m_str`, which since it's the same address as the moved-to object also invalidates it's buffer. A subsequent resize then crashes. I assume the move ctor/assignment is called from the `std::map<> docEntries` inside `SearchIndexExternal::Private` when adding/removing entries.

Different approaches exist to implement move ctors/assignment ops. I'm personally a fan of using `std::exchange()` in such occasions, e.g.:
```cpp
    GrowBuf(GrowBuf &&other) 
      : m_str(std::exchange(other.m_str, 0))
      , m_pos(std::exchange(other.m_pos, 0))
      , m_len(std::exchange(other.m_len, 0))
    {}
    GrowBuf &operator=(GrowBuf &&other)
    {
      if (this == &other) 
        return *this;
      free(m_str);
      m_str = std::exchange(other.m_str, 0);
      m_pos = std::exchange(other.m_pos, 0);
      m_len = std::exchange(other.m_len, 0);
      return *this;
    }
```

As I mentioned earlier I have troubles getting FLEX and BISON deployed on my machine so I can't build Doxygen myself and directly create a PR. Let me know if you have further questions, I'm happy to help out.

**Comment by albert-github** [comment]
The mentioned test case crashes at my place under Windows and Cygwin as well with the mentioned version.

Indeed flex/ bison can sometimes be a bit troublesome to install.
- In Github Actions we use `choco install winflexbison` but this implies of course that `choco` is installed (I have no experience with installing through `choco`). See .github/workflows/build_cmake.yml
- in the manual a link to https://sourceforge.net/projects/winflexbison/is mentioned, did you try this (you need / best is to rename `win_flex.exe` to `flex.exe` and `winbison.exe` to `bison.exe`). Furthermore I added the directory where the flex / bison e3xecutables reside to the `PATH`. This works well for me (Note: I use `nmake` to build, not the projectfiles etc.)

**Comment by doxygen** [comment]
@Schulteatq I was able to reproduce this issue with your example, and your analysis of the move constructor is correct. It should at least initialize the `m_str` of `other` to 0 to prevent a double free. Not sure why I overlooked that one. I've now took over your suggestion, and for me it fixes the problem. Thanks for the help.

**Comment by albert-github** [comment]
I just tested it with Windows and Cygwin and the given example runs.

**Comment by Schulteatq** [comment]
I just tested your commit on a large part of our code base and everything seems to work fine. 

Thank you so much for your continuous efforts in maintaining Doxygen!

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.3.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
